filealreadyexist exception when run on quarkus guides

i assume quarkus current guides breaks some assumptions but might be good to handle the error better.

here is what happens when run on quarkus/asciidoc guides

```
 java -jar ~/code/work/asciidoc-splitter/target/asciidoc-splitter-1.0-SNAPSHOT.jar -s . -o ../result
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.headius.backport9.modules.Modules (file:/Users/max/code/work/asciidoc-splitter/target/asciidoc-splitter-1.0-SNAPSHOT.jar) to method sun.nio.ch.NativeThread.signal(long)
WARNING: Please consider reporting this to the maintainers of com.headius.backport9.modules.Modules
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
java.lang.RuntimeException: java.nio.file.FileAlreadyExistsException: /Users/max/code/work/quarkus/quarkus/docs/src/main/asciidoc/../result/modules/native-limitations.adoc
	at com.redhat.documentation.asciidoc.cli.ExtractionRunner.writeModules(ExtractionRunner.java:155)
	at com.redhat.documentation.asciidoc.cli.ExtractionRunner.call(ExtractionRunner.java:75)
	at com.redhat.documentation.asciidoc.cli.ExtractionRunner.call(ExtractionRunner.java:38)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1783)
	at picocli.CommandLine.access$900(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2150)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2144)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2108)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:1975)
	at picocli.CommandLine.execute(CommandLine.java:1904)
	at com.redhat.documentation.asciidoc.cli.ExtractionRunner.main(ExtractionRunner.java:56)
Caused by: java.nio.file.FileAlreadyExistsException: /Users/max/code/work/quarkus/quarkus/docs/src/main/asciidoc/../result/modules/native-limitations.adoc
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:94)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:215)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:370)
	at java.base/java.nio.file.Files.createFile(Files.java:647)
	at com.redhat.documentation.asciidoc.cli.ExtractionRunner.writeModules(ExtractionRunner.java:124)
	... 10 more
```

﻿

--- a/src/main/java/com/redhat/documentation/asciidoc/Util.java
+++ b/src/main/java/com/redhat/documentation/asciidoc/Util.java
@@ -1,0 +1,17 @@
+package com.redhat.documentation.asciidoc;
+
+import org.asciidoctor.ast.ContentNode;
+
+public class Util {
+
+        static public String getFullId(ContentNode node) {
+            StringBuilder buf = new StringBuilder(node.getId());
+
+            while(node.getParent()!=null) {
+                buf.insert(0, node.getId() + "/");
+                node = node.getParent();
+            }
+
+            return buf.toString();
+        }
+}

--- a/src/main/java/com/redhat/documentation/asciidoc/cli/ExtractionRunner.java
+++ b/src/main/java/com/redhat/documentation/asciidoc/cli/ExtractionRunner.java
@@ -1,46 +1,32 @@
 package com.redhat.documentation.asciidoc.cli;
 
-import java.io.BufferedReader;
-import java.io.File;
-import java.io.FileWriter;
-import java.io.IOException;
-import java.io.InputStreamReader;
-import java.io.Writer;
-import java.net.URISyntaxException;
-import java.nio.file.FileAlreadyExistsException;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Objects;
-import java.util.Set;
-import java.util.concurrent.Callable;
-import java.util.regex.Pattern;
-import java.util.stream.Collectors;
-
 import com.redhat.documentation.asciidoc.Configuration;
 import com.redhat.documentation.asciidoc.extraction.Assembly;
 import com.redhat.documentation.asciidoc.extraction.ExtractedModule;
-import com.redhat.documentation.asciidoc.extraction.SourceExtractor;
 import org.asciidoctor.Asciidoctor;
 import org.asciidoctor.OptionsBuilder;
-import org.asciidoctor.ast.Block;
 import org.asciidoctor.ast.Document;
 import org.asciidoctor.ast.Section;
-import org.asciidoctor.ast.StructuralNode;
 import org.asciidoctor.jruby.AsciiDocDirectoryWalker;
 import picocli.CommandLine;
 import picocli.CommandLine.Command;
 import picocli.CommandLine.Option;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
 
 @Command(name = "extract", mixinStandardHelpOptions = true, version = "1.0",
          description = "Create a modular documentation layout from a directory of asciidoc files.")
 public class ExtractionRunner implements Callable<Integer> {
     private List<Assembly> assemblies;
     private Set<ExtractedModule> modules;
+    private List<Issue> issues = new ArrayList<>();
 
     @Option(names = {"-s", "--sourceDir"}, description = "Directory containing the input asciidoc files.", required = true)
     File inputDir;
@@ -77,9 +63,14 @@ public class ExtractionRunner implements Callable<Integer> {
             writeAssemblies(config);
         }
 
+        long errors = this.issues.stream().filter(Issue::isError).count();
+
+        System.out.println("Found " + this.issues.size() + " issues. " + errors + " Errors.");
+
         return 0;
     }
 
+    /** returns true if all went well. false if there was some problem with uniqueness. */
     private void findSections(Document doc) {
         // TODO: What should we do with preamble?
         // TODO: This should probably be configurable
@@ -87,10 +78,23 @@ public class ExtractionRunner implements Callable<Integer> {
             if (node instanceof Section && node.getLevel() == 1) {
                 var assembly = new Assembly((Section) node);
                 this.assemblies.add(assembly);
-                this.modules.addAll(assembly.getModules());
+                for (var module : assembly.getModules()) {
+                    if (!this.modules.add(module)) {
+                        var duplicate = this.modules.stream().filter(m -> {
+                            return m.equals(module);
+                        }).findFirst();
+
+                        addIssue(Issue.error("Module with non-unique id. " + module + " is a duplicate of " + duplicate, node));
+                    }
+                }
             }
         });
         // We should have all the assemblies, and all of their modules now
+    }
+
+    private void addIssue(Issue error) {
+        System.out.println(error);
+        this.issues.add(error);
     }
 
     private void writeAssemblies(Configuration config) {
@@ -117,12 +121,22 @@ public class ExtractionRunner implements Callable<Integer> {
         // Create the modules directory and write the files
         try {
             // Create the output directories
-            Path modulesDir = Files.createDirectories(Paths.get(config.getOutputDirectory().getAbsolutePath(),
-                    "modules"));
+            Path modulesDir = Files.createDirectories(config.getOutputDirectory().toPath().resolve("modules"));
+            Set<Path> visitedPaths = new HashSet<>();
 
             for (ExtractedModule module : this.modules) {
                 // Create output file
-                Path moduleOutputFile = Files.createFile(Paths.get(modulesDir.toString(), module.getFileName()));
+                Path moduleOutputFile = Paths.get(modulesDir.toString(), module.getFileName());
+
+                if(moduleOutputFile.toFile().exists()) {
+                    if(visitedPaths.contains(moduleOutputFile)) {
+                        System.err.println("Already written to this file: " + moduleOutputFile + " for " + module);
+                        return;
+                    }
+                } else {
+                    moduleOutputFile = Files.createFile(moduleOutputFile);
+                }
+                visitedPaths.add(moduleOutputFile);
 
                 // Output the module
                 try (Writer output = new FileWriter(moduleOutputFile.toFile())) {
@@ -151,10 +165,6 @@ public class ExtractionRunner implements Callable<Integer> {
                 }
             }
         } catch (IOException e) {
-            if (e instanceof FileAlreadyExistsException) {
-                System.err.println("Could not write a file: " + ((FileAlreadyExistsException) e).getFile());
-                return;
-            }
             // TODO: We blew-up in an unexpected way, handle this
             throw new RuntimeException(e);
         }

--- a/src/main/java/com/redhat/documentation/asciidoc/cli/Issue.java
+++ b/src/main/java/com/redhat/documentation/asciidoc/cli/Issue.java
@@ -1,0 +1,44 @@
+package com.redhat.documentation.asciidoc.cli;
+
+import org.asciidoctor.ast.StructuralNode;
+
+/**
+ * Represents an issue (warning/error) found in the processing of the documents.
+ */
+public class Issue {
+
+    private final StructuralNode node;
+    private final boolean error;
+    private final String text;
+
+    static Issue error(String text, StructuralNode node) {
+        return new Issue(true, text, node);
+    }
+
+    static Issue nonerror(String text, StructuralNode node) {
+        return new Issue(false, text, node);
+    }
+
+    Issue(boolean error, String text, StructuralNode node) {
+        this.error = error;
+        this.text = text;
+        this.node = node;
+    }
+
+
+    boolean isError() {
+        return error;
+    }
+
+    @Override
+    public String toString() {
+        String contextinfo = null;
+        if(node.getSourceLocation()==null) {
+            contextinfo = node.getClass().toString();
+        } else {
+            contextinfo = node.getSourceLocation().toString();
+        }
+
+        return (error ? "ERROR: " : "WARNING: ") + contextinfo + ": " + text;
+    }
+}

--- a/src/main/java/com/redhat/documentation/asciidoc/extraction/ExtractedModule.java
+++ b/src/main/java/com/redhat/documentation/asciidoc/extraction/ExtractedModule.java
@@ -5,12 +5,23 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 
+import com.redhat.documentation.asciidoc.Util;
 import org.asciidoctor.ast.Section;
 
 public class ExtractedModule {
     private String id;
     private Section section;
     private List<String> sources;
+
+    @Override
+    public String toString() {
+        return "ExtractedModule{" +
+                "id='" + id + '\'' +
+                //", section=" + section +
+                // ", sources=" + sources +
+                ", parentid=" + Util.getFullId(section.getParent()) +
+                '}';
+    }
 
     public ExtractedModule(Section section) {
         // According to the modular docs, there should only be one underscore used to split the context.

